### ISSUE TITLE
Diagnostic Translation

### ISSUE THREAD

**Issue by davidtwco** [rust-lang/rust#100717]
The Rust Diagnostics working group is leading an effort to add support for internationalization of error messages in the compiler, allowing the compiler to produce output in languages other than English. This issue tracks the current status of the effort, which was announced in the [*"Contribute to the diagnostic translation effort!"* post on *Inside Rust*](https://blog.rust-lang.org/inside-rust/2022/08/16/diagnostic-effort.html).

## What's the current status?
Diagnostic translation will take a long time to be finished. At a high-level, there are four primary steps: 

- [x] Implement initial translation infrastructure
- [ ] Make diagnostics translatable through migration to new infrastructure (**we are here**)
- [ ] Set up Pontoon for translators to use
- [ ] Establish translation teams for different languages
- [ ] Implement infrastructure for distributing language packs in collaboration with infrastructure/release teams (as appropriate)

Implementing the initial translation infrastructure provides the groundwork that enables diagnostic messages to be made translatable at all. That initial infrastructure is largely completed - there might be some gaps that we'll discover and patch up as we continue - but it's almost all there.

Next, all of the diagnostics in rustc need to be modified so that they can be translatable. There's some bad news - that's a _lot_ of work. But there's also some good news - that's a lot of _highly parallelizable_ work that you can help with! It doesn't require any familiarity with the Rust compiler, just an eagerness to get involved.

## How to get started?
It's very easy to get started, the process looks like the following:

1. Join [our Zulip chat](https://rust-lang.zulipchat.com/#narrow/stream/147480-t-compiler.2Fwg-diagnostics) and say hello! Everyone is very friendly and eager to help if you have any trouble.
1. [Set up a development environment](https://rustc-dev-guide.rust-lang.org/building/how-to-build-and-run.html) for the compiler.
1. Identify a module to migrate (see ###).
1. Migrate diagnostics (see ###).
1. Open a pull request with your changes.
1. Repeat and profit!

### Identifying diagnostics to migrate
Our goal is to migrate every diagnostic in the compiler to be translatable and to switch from using a "diagnostic builder" to using "diagnostic structs". That's a lot of diagnostics, so we're splitting the work up by module in the compiler so that nobody steps on anyone else's toes.

- [ ] `rustc_apfloat` (currently being worked on by `@nobody`)
- [ ] `rustc_arena` (currently being worked on by `@nobody`)
- [ ] `rustc_ast` (currently being worked on by `@nobody`)
- [ ] `rustc_ast_lowering` (currently being worked on by `@nobody`)
- [ ] `rustc_ast_passes` (currently being worked on by `@nobody`)
- [ ] `rustc_ast_pretty` (currently being worked on by `@nobody`)
- [ ] `rustc_attr` (currently being worked on by `@nobody`)
- [ ] `rustc_borrowck` (currently being worked on by `@nobody`)
- [ ] `rustc_builtin_macros` (currently being worked on by `@nobody`)
- [ ] `rustc_codegen_cranelift` (currently being worked on by `@nobody`)
- [ ] `rustc_codegen_gcc` (currently being worked on by `@nobody`)
- [ ] `rustc_codegen_llvm` (currently being worked on by `@nobody`)
- [ ] `rustc_codegen_ssa` (currently being worked on by `@nobody`)
- [ ] `rustc_const_eval` (currently being worked on by `@nobody`)
- [ ] `rustc_data_structures` (currently being worked on by `@nobody`)
- [ ] `rustc_driver` (currently being worked on by `@nobody`)
- [ ] `rustc_error_codes` (currently being worked on by `@nobody`)
- [ ] `rustc_error_messages` (currently being worked on by `@nobody`)
- [ ] `rustc_errors` (currently being worked on by `@nobody`)
- [ ] `rustc_expand` (currently being worked on by `@nobody`)
- [ ] `rustc_feature` (currently being worked on by `@nobody`)
- [ ] `rustc_fs_util` (currently being worked on by `@nobody`)
- [ ] `rustc_graphviz` (currently being worked on by `@nobody`)
- [ ] `rustc_hir` (currently being worked on by `@nobody`)
- [ ] `rustc_hir_pretty` (currently being worked on by `@nobody`)
- [ ] `rustc_incremental` (currently being worked on by `@nobody`)
- [ ] `rustc_index` (currently being worked on by `@nobody`)
- [ ] `rustc_infer` (currently being worked on by `@nobody`)
- [ ] `rustc_interface` (currently being worked on by `@nobody`)
- [ ] `rustc_lexer` (currently being worked on by `@nobody`)
- [ ] `rustc_lint` (currently being worked on by `@nobody`)
- [ ] `rustc_lint_defs` (currently being worked on by `@nobody`)
- [ ] `rustc_llvm` (currently being worked on by `@nobody`)
- [ ] `rustc_log` (currently being worked on by `@nobody`)
- [ ] `rustc_macros` (currently being worked on by `@nobody`)
- [ ] `rustc_metadata` (currently being worked on by `@nobody`)
- [ ] `rustc_middle` (currently being worked on by `@nobody`)
- [ ] `rustc_mir_build` (currently being worked on by `@nobody`)
- [ ] `rustc_mir_dataflow` (currently being worked on by `@nobody`)
- [ ] `rustc_monomorphize` (currently being worked on by `@nobody`)
- [ ] `rustc_parse` (currently being worked on by `@nobody`)
- [ ] `rustc_parse_format` (currently being worked on by `@nobody`)
- [ ] `rustc_passes` (currently being worked on by `@nobody`)
- [ ] `rustc_plugin_impl` (currently being worked on by `@nobody`)
- [x] `rustc_privacy` ([completed by `@davidtwco`](https://github.com/rust-lang/rust/blob/361c599feeefaf6e50efd90658fc9c2222154684/compiler/rustc_privacy/src/lib.rs#L8-L9))
- [ ] `rustc_query_impl` (currently being worked on by `@nobody`)
- [ ] `rustc_query_system` (currently being worked on by `@nobody`)
- [ ] `rustc_resolve` (currently being worked on by `@nobody`)
- [ ] `rustc_save_analysis` (currently being worked on by `@nobody`)
- [ ] `rustc_serialize` (currently being worked on by `@nobody`)
- [ ] `rustc_session` (currently being worked on by `@nobody`)
- [ ] `rustc_smir` (currently being worked on by `@nobody`)
- [ ] `rustc_span` (currently being worked on by `@nobody`)
- [ ] `rustc_symbol_mangling` (currently being worked on by `@nobody`)
- [ ] `rustc_target` (currently being worked on by `@nobody`)
- [ ] `rustc_trait_selection` (currently being worked on by `@nobody`)
- [ ] `rustc_traits` (currently being worked on by `@nobody`)
- [ ] `rustc_transmute` (currently being worked on by `@nobody`)
- [ ] `rustc_ty_utils` (currently being worked on by `@nobody`)
- [ ] `rustc_type_ir` (currently being worked on by `@nobody`)
- [ ] `rustc_typeck` (currently being worked on by `@nobody`)

**Note:** Some of these crates might not have diagnostics in them, in which case we'll just enable our internal lints on them. Some might have lots and lots of work that we can split up further, let us know! If there aren't many crates left, then feel free to leave a comment asking if someone is still working on their crate (check if they commented or have put a PR up recently).

Once you've picked a module (**leave a comment letting us know!**), how do you find the diagnostics to migrate? We've created rustc-internal lints that you can apply to a module which will produce an error for every diagnostic that hasn't been migrated.

```rust=
#![deny(rustc::untranslatable_diagnostics)]
#![deny(rustc::diagnostic_outside_of_impl)]
```

(*an example of using these would just be adding them to the top of a file*)

After adding these attributes, you can run `./x.py check` to build the compiler in check mode (just like `cargo check` in another project). You'll notice a bunch of errors that will look something like these:

```text=
error: diagnostics should only be created in `SessionDiagnostic`/`AddSubdiagnostic` impls
    --> compiler/rustc_parse/src/parser/mod.rs:1443:40
     |
1443 |     let mut err = sess.span_diagnostic.struct_span_err(
     |                                        ^^^^^^^^^^^^^^^

error: diagnostics should be created using translatable messages
    --> compiler/rustc_parse/src/parser/mod.rs:1443:40
     |
1443 |     let mut err = sess.span_diagnostic.struct_span_err(
     |                                        ^^^^^^^^^^^^^^^
```

There will be two errors for each diagnostic that isn't migrated:

1. *"diagnostics should be created using translatable messages"*
    - This error occurs when a diagnostic function is being invoked with something that isn't a translatable message (like a string literal or a formatted string).
        - e.g. `err.label("an example label")` instead of `err.label(fluent::example_label)`
        - `fluent::example_label` corresponds to a message in a "Fluent resource" which we can provide different versions of for each language.
2. *"diagnostics should only be created in `SessionDiagnostic`/`AddSubdiagnostic` impls"*
    - This error occurs when a diagnostic function is being called outside of an impl of `SessionDiagnostic` or `AddSubdiagnostic`. One of our goals with this migration is to move all diagnostic emission logic into impls on structs, as it helps keep the compiler tidy and works towards other goals of the diagnostics working group.
    - There are two ways to resolve this:
        1. Using a diagnostic derive to implement them automatically (preferred!)
        1. Implementing one of these traits (`SessionDiagnostic` for errors and warnings, `LintDiagnostic` for lints, or `AddSubdiagnostic` for parts of an error/warning/lint) manually.
    - See ### for more on these.

We'll know we're finished when we can leave those attributes on every module in the compiler.

### Migrate diagnostics
Okay, so you've got a diagnostic in front of you that you need to migrate.. now what? 

- There's an introduction to performing a migration in the [*"Contribute to the diagnostic translation effort!"* post on *Inside Rust*](https://blog.rust-lang.org/inside-rust/2022/08/16/diagnostic-effort.html), this should serve as a decent introduction to the process.
- There's detailed documentation on [diagnostic structs](https://rustc-dev-guide.rust-lang.org/diagnostics/diagnostic-structs.html) and on [diagnostic translation](https://rustc-dev-guide.rust-lang.org/diagnostics/translation.html) in the developer guide that should be useful reference material.
- There are a lot of pull requests that perform migrations that you can dig through for examples, just [look for pull requests labelled `A-translation`](https://github.com/rust-lang/rust/pulls?q=is%3Apr+is%3Aopen+sort%3Aupdated-desc+label%3AA-translation).

While migrating diagnostics, there might be cases you run into that we've not run across yet. Let us know in Zulip, you might be able to experiment and teach us how to translate some diagnostics, or there may be an opportunity to extend our core infrastructure (e.g. the derives). Don't worry though, you can always skip a diagnostic and leave it for someone else too.

## Where to get help?
Discussion is primarily happening in the [`#wg-diagnostics` stream on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/147480-t-compiler.2Fwg-diagnostics). Ask any questions you have in that chat and someone will try to help. If you don't get a response, feel free to ping `@davidtwco` or `@t-compiler/wg-diagnostics`.

**Comment by nidnogg** [comment]
I took up `rustc_const_eval`  and `rustc_expand` for the time being. Kudos for the further write-up! 🚀

**Comment by JeanCASPAR** [comment]
I'm migrating `rustc_ast_lowering`

**Comment by IntQuant** [comment]
I'll probably look into migrating `rustc_infer`

**Comment by Xiretza** [comment]
I'm working on `rustc_parse`.

**Comment by 5225225** [comment]
I'm doing the zero diagnostic crates in one PR, will constantly edit this comment to keep track of which ones they were.

Currently: `rustc_apfloat`, `rustc_arena`, `rustc_traits`

**Comment by superblaubeere27** [comment]
I'm working on `rustc_typeck`.

**Comment by JhonnyBillM** [comment]
Working on `rustc_type_ir` (zero diagnostic)
PR: #100721

**Comment by CleanCut** [comment]
@5225225  `rustc_log` seems to be zero diagnostics as well. Do you want to add that to your PR?

I'll continue to look for something unclaimed and small...

**Comment by CleanCut** [comment]
> [ ] `rustc_transmute` (currently being worked on by `@nobody`)

 🤷🏻‍♂️ I can't find this one. Should I be on a branch other than `master`?

**Comment by 5225225** [comment]
Update your checkout. It's on `master`, but I think it's fairly new. (I believe it's the safe transmute people)

**Comment by CleanCut** [comment]
> Update your checkout. It's on `master`, but I think it's fairly new. (I believe it's the safe transmute people)

🤦🏻‍♂️ Thanks! I was working off a clone of my fork. I'm glad I discovered this now.

**Comment by 5225225** [comment]
Also if you want something small, rustc_monomorphize has 16 errors.

**Comment by CleanCut** [comment]
> Also if you want something small, rustc_monomorphize has 16 errors.

Working on `rustc_monomorphize`

**Comment by finalchild** [comment]
Working on `rustc_ast_passes`

**Comment by Facel3ss1** [comment]
Working on `rustc_ty_utils`

**Comment by compiler-errors** [comment]
I marked off all of the crates in @5225225's #100723

**Comment by CleanCut** [comment]
I can't figure out how to do an optional `#[note = "..."]`. If someone could take a look at [my draft PR](https://github.com/rust-lang/rust/pull/100730) and steer me in the right direction, that'd be most helpful. :smile:

**Comment by Rejyr** [comment]
Working on `rustc_lint`

**Comment by gabrielBusta** [comment]
Working on `rustc_trait_selection`

**Comment by evopen** [comment]
working on `rustc_query_system`

**Comment by AndyJado** [comment]
working on `rustc_borrowck`

**Comment by LuisCardosoOliveira** [comment]
Working on `rustc_session`

**Comment by 5225225** [comment]
Working on `rustc_mir_dataflow`

**Comment by davidtwco** [comment]
It's so exciting to wake up to all this interest and progress, I've updated the list with what everyone is working on.

**Comment by hampuslidin** [comment]
I would like to work on `rustc_attr`.

**Comment by davidtwco** [comment]
A bonus task for someone who is interested, @compiler-errors makes the good suggestion that we could replace `emit_err`, `emit_warn` and `emit_fatal` (after #100694) with a single function which is generic over the `EmissionGuarantee`-impl'ing type that separates these different kinds of diagnostics.

> Hm, maybe we should expose some `sess.struct_$SOMETHING` (like `struct_diagnostic`?) that is generic over `EmissionGuarantee`, then make the `SessionDiagnostic` derive generic, i.e.
>  
> ```rust
> impl<'tcx> SessionDiagnostic for UnusedGenericParams {
>   fn into_diagnostic<T: EmissionGuarantee>( .. ) -> DiagnosticBuilder<'tcx, T> {
>     let mut diag = sess.struct_diagnostic(rustc_errors:..);
>     ..
>   }
> }
> ```
> 
> Like, ideally we shouldn't care what kind of `DiagnosticBuilder` (i.e. what kind of emission) we want when designing this struct. We should be able to do `tcx.emit_err(Foo);` or `tcx.emit_warn(Foo);` etc.

(https://github.com/rust-lang/rust/pull/100730#discussion_r949712629)

**Comment by davidtwco** [comment]
Another small task that someone could take on - we have `LintDiagnostic` which shares a lot of the same infrastructure as `SessionDiagnostic`. `LintDiagnostic` structs shouldn't use `#[primary_span]` because they are invoked through [`emit_spanned_lint`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/context/struct.TyCtxt.html#method.emit_spanned_lint) which takes the primary `Span` as an argument (this is necessary because we use that `Span` to calculate whether the lint at that location should be a warning or error - based on `allow`/`deny` attributes on parent items). It would be great if someone made it so that `LintDiagnostic` produces an error if you try use `#[primary_span]` in it.

**Comment by Xiretza** [comment]
> A bonus task for someone who is interested, @compiler-errors makes the good suggestion that we could replace `emit_err`, `emit_warn` and `emit_fatal` (after #100694) with a single function which is generic over the `EmissionGuarantee`-impl'ing type that separates these different kinds of diagnostics.

I'll look into that - I guess that would mean replacing the `#[error(...)]`/`#[warning(...)]` struct attributes with a generic `#[message(...)]` or somesuch?

**Comment by davidtwco** [comment]
> > A bonus task for someone who is interested, @compiler-errors makes the good suggestion that we could replace `emit_err`, `emit_warn` and `emit_fatal` (after #100694) with a single function which is generic over the `EmissionGuarantee`-impl'ing type that separates these different kinds of diagnostics.
> 
> I'll look into that - I guess that would mean replacing the `#[error(...)]`/`#[warning(...)]` struct attributes with a generic `#[message(...)]` or somesuch?

Not quite - we'd keep the `#[error(..)]`/`#[warning(..)]` as they are. Those will let us generate `SessionDiagnostic` impls with the appropriate `EmissionGuarantee` type, so `#[error(..)]` generates a `impl SessionDiagnostic<'_, ErrorGuaranteed> for Foo`, and `#[warning(..)]` generates a `impl SessionDiagnostic<'_, ()> for Foo` (and `#[fatal(..)]` is similar, but with a `!`). Instead of using `emit_err` and `emit_warning` and similar functions to *use* these types, we'd just have one function that could take any of them.

**Comment by Xiretza** [comment]
> Another small task that someone could take on - we have `LintDiagnostic` which shares a lot of the same infrastructure as `SessionDiagnostic`. `LintDiagnostic` structs shouldn't use `#[primary_span]` because they are invoked through [`emit_spanned_lint`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/context/struct.TyCtxt.html#method.emit_spanned_lint) which takes the primary `Span` as an argument (this is necessary because we use that `Span` to calculate whether the lint at that location should be a warning or error - based on `allow`/`deny` attributes on parent items). It would be great if someone made it so that `LintDiagnostic` produces an error if you try use `#[primary_span]` in it.

Since I'm working on that code anyway, I'll pick this up as well.

**Comment by Facel3ss1** [comment]
Working on `rustc_plugin_impl`

**Comment by rajputrajat** [comment]
started working on `rustc_resolve`

**Comment by wonchulee** [comment]
I'm working on `rustc_save_analysis`

**Comment by wonchulee** [comment]
started working on `rustc_builtin_macros`

**Comment by davidtwco** [comment]
Another day of great progress, really excited to see all the pull requests and activity!

**Comment by TheOddGarlic** [comment]
I'm working on `rustc_mir_build`

**Comment by SkiFire13** [comment]
I'm working on `rustc_interface`

**Comment by adriantombu** [comment]
I'll have a try at `rustc_driver`

**Comment by rdvdev2** [comment]
Started working on `rustc_passes`

**Comment by JhonnyBillM** [comment]
Working on `rustc_codegen_ssa` while I understand better how to proceed on [how to migrate rustc_errors](https://rust-lang.zulipchat.com/#narrow/stream/336883-i18n/topic/.23100717.20migrating.20rustc_errors). 
I can un-assign any of them if needed or desired.

Also from [my previous comment](https://github.com/rust-lang/rust/issues/100717#issuecomment-1219761885), working on:
- `rustc_transmute` _(zero diagnostic)_ - PR: #100842
- `rustc_errors`

**Comment by davidtwco** [comment]
Okay, I've tried to go through and update the status of each crate and link to any pull requests - thanks everyone for your contributions so far, we're making great progress. Feel free to ask in `#i18n` in Zulip if you have any questions, or your crate is really large and you could use a hand.

- @JeanCASPAR, now that fatal support has landed, it's my understanding that the remaining parts of `rustc_ast_lowering` can be migrated? Would you be up for finishing that crate off?
- @finalchild, is `rustc_ast_passes` complete after #100694 or is there more to do? If it is complete, could you open a PR adding the `#![deny(..)]` attributes for our migration lints?

Let me know if I've missed anything in the status list.

**Comment by CleanCut** [comment]
I'll take on `rustc_metadata`

**Comment by SLASHLogin** [comment]
I'll start working on `rustc_codegen_llvm`

**Comment by davidtwco** [comment]
Some one-off tasks for anyone interested, some quite large and involved, others very small:

- `SessionDiagnostic` on enums ([source](https://github.com/rust-lang/rust/pull/100831#discussion_r951403293))
- Renaming `warn_` on `SessionDiagnostic` to `warning` now that we use `diag` instead of `warning`.
- In [#100753](https://github.com/rust-lang/rust/pull/100753/files#diff-75e151b9de9e9418f72fab3e76335c676cbca4e50db984c2ddb63bdf1b7db3e3R375-R377) there are some functions annotated with `#[rustc_lint_diagnostics]` which is used to know when to trigger our internal lints, that are themselves triggering the internal lint - it would be good to change the internal lint so that it skips functions annotated with `#[rustc_lint_diagnostics]`.

**Comment by LuisCardosoOliveira** [comment]
> Some one-off tasks for anyone interested, some quite large and involved, others very small:
> 
> * `SessionDiagnostic` on enums ([source](https://github.com/rust-lang/rust/pull/100831#discussion_r951403293))
> * Renaming `warn_` on `SessionDiagnostic` to `warning` now that we use `diag` instead of `warning`.
> * In [#100753](https://github.com/rust-lang/rust/pull/100753/files#diff-75e151b9de9e9418f72fab3e76335c676cbca4e50db984c2ddb63bdf1b7db3e3R375-R377) there are some functions annotated with `#[rustc_lint_diagnostics]` which is used to know when to trigger our internal lints, that are themselves triggering the internal lint - it would be good to change the internal lint so that it skips functions annotated with `#[rustc_lint_diagnostics]`.

Hey! I'm working on the renaming task, might send a PR soon.

**Comment by MingyuChen1** [comment]
I'll start working on ``rustc_middle``

**Comment by ellishg** [comment]
Now that I've finally managed to build `rustc_codegen_gcc`, I'd like to try working on it!

**Comment by davidtwco** [comment]
Thanks everyone for your continued hard work on this, it's much appreciated, we're still making good progress!

I've went through the crate list and updated the statuses on everything, let me know if I've forgotten anything.

- I'm going to unassign @finalchild from `rustc_ast_passes` (if you're around though and want to continue, just ping me), so someone who wants to finish that crate off can feel free to do so.
- I'll also unassign @wonchulee from `rustc_builtin_macros` - if you're working on this, let me know, but there's not been a PR or any updates, so I'll free this up for someone else unless I hear otherwise.
- @JhonnyBillM, you're assigned to `rustc_codegen_ssa`, are you still planning on working on this? I know that you're still actively working on/waiting on bors for some other pull requests :)
- Similarly, @JhonnyBillM, you're assigned to `rustc_errors`, I know that there were some blockers with this crate, so I'll leave it assigned to you because I don't know that we're going to be able to make progress there quite yet.
- @nidnogg, you're assigned to `rustc_expand` and have landed one PR for that crate, are you planning to continue working on it? Similarly with `rustc_const_eval`?
- @Xiretza, I know you've been doing some excellent infrastructure work, are you still planning to continue with `rustc_parse`?
- @rdvdev2, are you still working on #100870? - there's no rush, there's just not been any updates, someone else might be able to pick up the work if you cannot.
- @superblaubeere27, @gabrielBusta, @MingyuChen1 (and anyone else this applies to), you're working on crates that are quite large, let me know if you want to split them up so others can help!

For anyone looking to get involved, there's still plenty to do - pick a unassigned crate above or one of the one-off tasks!

**Comment by Xiretza** [comment]
Still working on `rustc_parse`, I just got sidetracked repeatedly by infrastructure hurdles that I've had to cross while doing that :smile:

**Comment by JhonnyBillM** [comment]
@davidtwco 
1. Yes, I plan to open a PR this week for `rustc_codegen_ssa`, [here is the branch with progress](https://github.com/jhonnybillm/rust/tree/migrate-codegen-ssa-to-diagnostics-structs). In https://github.com/JhonnyBillM/rust/commit/7a41fb285ac19f7608f82ccebe69278254935caf I added a FIXME note that we will probably have to discuss.
2. Yes, I still plan to work on `rustc_errors`, but haven't made progress in the last few days. I do expect to focus a little bit more over the next few days once I am done with work and codegen_ssa.

**Comment by LuisCardosoOliveira** [comment]
> Thanks everyone for your continued hard work on this, it's much appreciated, we're still making good progress!
> 
> 
> 
> I've went through the crate list and updated the statuses on everything, let me know if I've forgotten anything.
> 
> 
> 
> - ~~I'm going to unassign @finalchild from `rustc_ast_passes` (if you're around though and want to continue, just ping me), so someone who wants to finish that crate off can feel free to do so.~~ https://github.com/rust-lang/rust/issues/100717#issuecomment-1232943107
> 
> - I'll also unassign @wonchulee from `rustc_builtin_macros` - if you're working on this, let me know, but there's not been a PR or any updates, so I'll free this up for someone else unless I hear otherwise.
> 
> - ~~@JhonnyBillM, you're assigned to `rustc_codegen_ssa`, are you still planning on working on this? I know that you're still actively working on/waiting on bors for some other pull requests :)~~ https://github.com/rust-lang/rust/issues/100717#issuecomment-1232939993
> 
> - ~~Similarly, @JhonnyBillM, you're assigned to `rustc_errors`, I know that there were some blockers with this crate, so I'll leave it assigned to you because I don't know that we're going to be able to make progress there quite yet.~~ https://github.com/rust-lang/rust/issues/100717#issuecomment-1232939993
> 
> - @nidnogg, you're assigned to `rustc_expand` and have landed one PR for that crate, are you planning to continue working on it? Similarly with `rustc_const_eval`?
> 
> - ~~@Xiretza, I know you've been doing some excellent infrastructure work, are you still planning to continue with `rustc_parse`?~~ https://github.com/rust-lang/rust/issues/100717#issuecomment-1232928039
> 
> - @rdvdev2, are you still working on #100870? - there's no rush, there's just not been any updates, someone else might be able to pick up the work if you cannot.
> 
> - @superblaubeere27, @gabrielBusta, @MingyuChen1 (and anyone else this applies to), you're working on crates that are quite large, let me know if you want to split them up so others can help!
> 
> 
> 
> For anyone looking to get involved, there's still plenty to do - pick a unassigned crate above or one of the one-off tasks!

It's so cool to see all this progress!

I'll be finishing `rustc_session` today or tomorrow first thing in the morning and I'll be avaliable to take another task!

BTW, thank you so much @davidtwco and all the diagnostic team! You folks have been really helpful and this project rocks ☀️☀️

**Comment by gabrielBusta** [comment]
@davidtwco 
> ... you're working on crates that are quite large, let me know if you want to split them up so others can help!

`rustc_trait_selection` is quite large indeed. I could definitely use some help.

[`traits/error_reporting/suggestions.rs`](https://github.com/rust-lang/rust/blob/15e2e5185a22207b18d2cbc47a48b39e63e84cd0/compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs) has a higher number of errors than the rest of the files added together.

`suggestions.rs` and "the rest of the crate" might be a good way to split up `rustc_trait_selection`

I've started to look at some of the errors in [`traits/error_reporting/mod.rs`](https://github.com/rust-lang/rust/blob/8189a4536be3730ecc161c77fc655b08be179b50/compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs)

We can probably split up the crate further into `suggestions.rs`, `error_reporting/mod.rs`, and "the rest of the crate"

**Comment by rdvdev2** [comment]
> * @rdvdev2, are you still working on https://github.com/rust-lang/rust/pull/100870? - there's no rush, there's just not been any updates, someone else might be able to pick up the work if you cannot.

Life got in the way, but now I'm retaking the work in the PR. I'm probably going to do some more commits the next hours.

**Comment by davidtwco** [comment]
> `rustc_trait_selection` is quite large indeed. I could use some help.
> 
> [`traits/error_reporting/suggestions.rs`](https://github.com/rust-lang/rust/blob/15e2e5185a22207b18d2cbc47a48b39e63e84cd0/compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs) has a slightly higher number of errors than the rest of the files added together.
> 
> We could split `rustc_trait_selection` into `suggestions.rs` and "the rest of the crate". We could also split it up further into `suggestions.rs`, `error_reporting/mod.rs`, and "the rest of the crate".
> 
> I started to look at some of the errors in [`traits/error_reporting/mod.rs`](https://github.com/rust-lang/rust/blob/8189a4536be3730ecc161c77fc655b08be179b50/compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs)

I've split this crate up as you suggested, `rustc_trait_selection::traits::error_reporting::suggestions` and everything else, and left your assignment on "everything else" given what you've said you're working on currently.

**Comment by wonchulee** [comment]
@davidtwco 

I'm still working on `rustc_builtin_macros` and will submit a pr shortly.

**Comment by str4d** [comment]
Working on `rustc_codegen_cranelift`.

**Comment by str4d** [comment]
@wonchulee whoops, I had searched the history of this PR for `rustc_codegen_cranelift` to check no one was working on it, but forgot to do the same when I switched to `rustc_builtin_macros`. I've opened #101408 with the work I've done so far to migrate (almost all of) `rustc_builtin_macros::asm`; I'll leave it there and move to something else.

**Comment by str4d** [comment]
Working on `rustc_trait_selection::traits::error_reporting::suggestions`.

**Comment by pierwill** [comment]
I can help with `rustc_middle`!

**Comment by finalchild** [comment]
Working on `rustc_ast_passes`!

**Comment by TheOddGarlic** [comment]
I'm currently battling issues with my dev environment, and school has just started. If someone is interested in continuing my work on #100854 then I'm down for it, but I'm still working on it

**Comment by diegooliveira** [comment]
I made a pull request rustc_passes.

https://github.com/rust-lang/rust/pull/101815

**Comment by wonchulee** [comment]
I didn't have a lot of free time to work these days, so I couldn't finish `builtin_macros` all, and I uploaded what I've been working on a PR (https://github.com/rust-lang/rust/pull/100780). I don't want to obstruct the development of this change, so if there is someone who will continue the work, please take it over but I will continue to update from time to time until someone takes it.

**Comment by rdvdev2** [comment]
I have to abandon `rustc_passes` (more details in #100870). Feel free to continue to work on it.

**Comment by CleanCut** [comment]
> I have to abandon `rustc_passes` (more details in #100870). Feel free to continue to work on it.

I have opened #102110 to pick up the closed PR above. I would love for some collaborators to come and join me in fixing this one, as it is rather large. I outlined the TODO items in the PR description. Come join in!

**Comment by JhonnyBillM** [comment]
[PR#101558](https://github.com/rust-lang/rust/pull/101558) has been merged! Please make sure to rename the macros and traits in your branches, here is what we renamed here:

1. [Macro] `SessionDiagnostic`-> `Diagnostic`

2. [Trait] `SessionDiagnostic` -> `IntoDiagnostic`

3. `SessionSubdiagnostic` -> `Subdiagnostic`

4. `AddSubdiagnostic` -> `AddToDiagnostic`

❗️~~SessionDiagnostic~~ `Diagnostic` macro has been moved from `rustc_session` to `rustc_errors`

_If you encounter any issues, please reach out in [#100717 SessionDiagnostic on Handler](https://rust-lang.zulipchat.com/#narrow/stream/336883-i18n/topic/.23100717.20SessionDiagnostic.20on.20Handler)_

**Comment by davidtwco** [comment]
I'm still catching up on reviews, so if you're waiting on that then I'll get to that soon - please message me if you think I might have lost track of your PR and I'll take a look!

I've updated this issue and the current status of each of the crates:

- I've unassigned @wonchulee from `rustc_builtin_macros`, thanks for your work on it, if someone else wants to pick up #101408 then please do! @str4d might be interested in this having opened #101408.
- @nidnogg, are you still working on `rustc_const_eval` and `rustc_expand`?
- @pierwill, have you made any progess on `rustc_middle`?
- @SLASHLogin, what is the current status of #101005, I lost track of it - is it awaiting my review?
- @JhonnyBillM, I know that you've been putting in other pull requests, are you still intending to do `rustc_codegen_ssa`? Similarly, is `rustc_errors` more achievable now that the `Diagnostic` derive has moved?
- @gabrielBusta, are you still working on `rustc_trait_selection`?
- @superblaubeere27, are you still working on `rustc_typeck`?
- I've unassigned @rdvdev2 from `rustc_passes`, thanks for your work on it, and thanks to @CleanCut and @diegooliveira for picking that up.

**Comment by JhonnyBillM** [comment]
Yes to both `rustc_codegen_ssa` and `rustc_errors`. I was waiting for the move+rename PR to get in first. 
Plan to open a PR for `codegen_ssa` between today and tomorrow

**Comment by nidnogg** [comment]
Heya @davidtwco! Yes I'll be working on both `rustc_expand` and `rustc_const_eval` probably this week as I've recently finished a chunk of work that was holding me back. I might need help with some of the trickier migrations though, I'll beep the #i18n channel as needed.

**Comment by CleanCut** [comment]
https://github.com/rust-lang/rust/pull/102110 is ready for review!

**Comment by wonchulee** [comment]
> I didn't have a lot of free time to work these days, so I couldn't finish `builtin_macros` all, and I uploaded what I've been working on a PR (#101935). I don't want to obstruct the development of this change, so if there is someone who will continue the work, please take it over but I will continue to update from time to time until someone takes it. (cc. @str4d)

I mislinked work in progressed PR, fixed to point https://github.com/rust-lang/rust/pull/101935

**Comment by camsteffen** [comment]
`rustc_ty_utils` is incomplete after https://github.com/rust-lang/rust/pull/102622. I updated the OP.

**Comment by davidtwco** [comment]
Hey everyone - been a little behind but updated the issue again with the current status of everything.

- `rust_ast_passes` remains incomplete. @finalchild, are you going to be able to address the comments on #101657?
- `rustc_borrowck` is still in progress by @AndyJado
- `rustc_builtin_macros` remains incomplete. @wonchulee, are you going to be able to address the comments on #101935? I've marked it as available and we can change it back if you're going to continue.
- `rustc_codegen_cranelift` is blocked because it's hard.
- `rustc_codegen_gcc` was completed by @ellishg
- `rustc_codegen_llvm` is very nearly complete thanks to @SLASHLogin, just two more nitpicks of mine to resolve and then we can land that PR.
- `rustc_codegen_ssa` has made a bunch of progress thanks to @JhonnyBillM, it doesn't have the lints denying untranslatable diagnostics yet - is there more work to do here @JhonnyBillM?
- `rustc_const_eval` doesn't seem to be finished yet, are you planning to continue to work on it @nidnogg? I've marked it as available and we can change it back if you still are.
- `rustc_errors` doesn't seem to have had much progress - what was the status here @JhonnyBillM?
- `rustc_expand` doesn't seem to be finished yet either, are you planning to continue to work on it @nidnogg? I've marked it as available and we can change it back if you still are.
- `rustc_infer` is still being worked on by @IntQuant
- `rustc_lint` is still being worked on by @Rejyr
- `rustc_middle` is assigned to @pierwill, did you start working on this? I've unassigned you for now but can re-assign if you are going to revisit this.
- `rustc_mir_build` is partially completed in #100854, but @AsyaTheAbove doesn't have time to finish it (https://github.com/rust-lang/rust/pull/100854#issuecomment-1300481691), if someone wants to pick that up, it would be great.
- `rustc_parse` has seen progress but is still incomplete. @Xiretza, do you intend to continue working on it? I've marked it as available and can change that if you do.
- `rustc_query_system` was finished off now that eager translation has landed.
- `rustc_resolve` is still being worked on by @rajputrajat
- `rustc_trait_selection::traits::error_reporting::suggestions` remains incomplete. @str4d, are you going to be able to address the comments on #101466? I've marked it as available and we can change it back if you're going to continue.
- `rustc_trait_selection` isn't finished yet. @gabrielBusta, would you like to continue working on it? I've marked it as available and we can change it back if you're going to continue.
- `rustc_ty_utils` has some more migration to do after the diagnostic migration lints were removed.
- `rustc_typeck` has been renamed to `rustc_hir_typeck` and remains incomplete. @superblaubeere27, are you going to continue working on it? I've marked it as available and we can change it back if you're going to continue.

After this update, the following areas are available:

- `rustc_const_eval`
- `rustc_expand`
- `rustc_hir_analysis`
- `rustc_hir_typeck`
- `rustc_middle`
- `rustc_parse`
- `rustc_resolve`
- `rustc_trait_selection::traits::error_reporting::suggestions`
- `rustc_trait_selection` (everything else)
- `rustc_ty_utils`
- `rustfmt`
- `rustdoc`
- `clippy`

I've also added new areas to be worked on:

- `rustc_hir_analysis`
- `rustdoc`
- `clippy`
- `rustfmt`

If I've missed anything, let me know!

There's new one-off tasks for anyone who wants them:

- #103539 (currently assigned to @TaKO8Ki)
- Add a `IntoDiagnosticSpan` trait that can be implemented for anything usable with `#[primary_span]` (i.e. so we can extend `#[primary_span]` support to `Ident` easily, for example)

**Comment by adriantombu** [comment]
I can have a try at `rustc_resolve`

**Comment by Xiretza** [comment]
I still have a bit more work for `rustc_parse` in the pipeline, I'll get that submitted shortly.

**Comment by JhonnyBillM** [comment]
I am working on the 3rd (*and final*) part of `codegen_ssa`, but it depends on `rustc_middle::mir::interpret::InterpError` _and friends_ from `rustc_middle` / `rustc_ty_utils`. 
If those modules are available, I could take them as part of finishing `codegen_ssa`.

As for `rustc_errors`,  I've tried a few things but haven't gotten to a final solution yet. I think it would be better for someone to take this one as I finish my current tasks.

**Comment by pierwill** [comment]
> rustc_middle is assigned to @pierwill, did you start working on this? I've unassigned you for now but can re-assign if you are going to revisit this.

There are still a few errors left, but they're complex enough that I don't think I can handle them in the near future. I might play with one or two, but I'm happy to have someone else assigned if they'd like!

**Comment by Nilstrieb** [comment]
I'm gonna be working on `rustc_expand`

**Comment by davidtwco** [comment]
> I can have a try at `rustc_resolve`

I think I had that listed in the "available to work on" section incorrectly, there are other crates available.

**Comment by nidnogg** [comment]
> There are still a few errors left, but they're complex enough that I don't think I can handle them in the near future. I might play with one or two, but I'm happy to have someone else assigned if they'd like!
Regarding my tasks with `rustc_expand` and `rustc_const_eval`, I share the same feeling. I did pick it back up recently and had a tough time getting ticking off some migrations though. If I find some time down the road I'll check here or Zulip for any available crates to work on.

@Nilstrieb if you've made any progress on `rustc_expand`, I'll be looking forward to seeing your PR for it!

**Comment by adriantombu** [comment]
I'm starting `rustc_hir_analysis` today, it looks like no one has taken it yet :)

**Comment by mejrs** [comment]
I plan to continue migrating mir_build

**Comment by shogohida** [comment]
Hi @davidtwco, I just want to clarify one thing. At which stage will translation works begin? 

I read [this article](https://blog.rust-lang.org/inside-rust/2022/08/16/diagnostic-effort.html) and came to this issue. I wanted to do Japanese translations but I realized that the present status is "Make diagnostics translatable through migration to new infrastructure". 

I might not be able to help at this stage but I want to contribute for translations!

**Comment by cassaundra** [comment]
I'm working on migrating `rustdoc` now.

**Comment by davidtwco** [comment]
> Hi @davidtwco, I just want to clarify one thing. At which stage will translation works begin?

Hopefully soon, setting up the infrastructure for this to start happening is next on my to-do list.

**Comment by obeis** [comment]
Working on `rustc_hir_analysis`

**Comment by tshepang** [comment]
> I'm gonna be working on `rustc_expand`

I translated one message (#108436), unaware of your claim. Should I find something else?

**Comment by Nilstrieb** [comment]
@tshepang  I don't plan on working on it, feel free. I've already unassigned myself in the issue description

**Comment by tshepang** [comment]
are you able to add me on the description (as well as @obeis to `rustc_hir_analysis`)

**Comment by mattjperez** [comment]
gonna work on `rustc_hir_typck`

**Comment by clubby789** [comment]
I'm working on `rustc_builtin_macros`

**Comment by thaodt** [comment]
I'm gonna working on `rustc_middle`

**Comment by gibbyfree** [comment]
I would like to work on `rustc_mir_build`

**Comment by clubby789** [comment]
I completed migrating `rustc_passes` in #110816

**Comment by clubby789** [comment]
I've been working on a project to track the number/location of remaining untranslatable diagnostics, and I thought it might be interesting/useful to look at the volume of work needing to be done and where - https://arewe.translatable.yet.c7.rs/

**Comment by blyxyas** [comment]
I'm going to try and translate the whole `rustc_lint` module. If someone is also translating that module, please notify so we don't repeat work!

**Comment by Rejyr** [comment]
> I'm going to prepare and translate the whole `rustc_lint` module. If someone is also translating that module, please notify so we don't repeat work!

Hey @blyxyas,

I've worked on `rust_lint` previously, and it seems complete. You should check it anyway to see if there are any untranslated diagnostics.

cc @davidtwco 
Could you unassign me from `rustc_lint`?  I've been busy lately.

**Comment by IntQuant** [comment]
cc @davidtwco Could you unassign me from `rustc_infer`?
(Also #107034 and #108698 seem to be missing from the list)

**Comment by blyxyas** [comment]
@Rejyr Maybe I'm mistaking what an untranslated diagnostic is.
Would [this](https://github.com/rust-lang/rust/blob/27d22d2045478c544ad043556de4993eebe4be29/compiler/rustc_lint/src/builtin.rs#L2483C33-L2493) be an untranslated diagnostic? I feel like it is because, well, it's a string in English. But maybe I'm missinterpreting what a diagnostic is.

**Comment by Rejyr** [comment]
> @Rejyr Maybe I'm mistaking what an untranslated diagnostic is. Would [this](https://github.com/rust-lang/rust/blob/27d22d2045478c544ad043556de4993eebe4be29/compiler/rustc_lint/src/builtin.rs#L2483C33-L2493) be an untranslated diagnostic? I feel like it is because, well, it's a string in English. But maybe I'm missinterpreting what a diagnostic is.

Yes, that would be an untranslated diagnostic. I suppose I didn't look close enough (whoops). Feel free to start working on them and open a PR.

**Comment by clubby789** [comment]
`mir_transform` doesn't seem to be listed here, but I've migrated it in #111004

**Comment by blyxyas** [comment]
Ok, after seeing this for about a day, I have no idea how to proceed.
`rustc_lint` is free.

**Comment by fee1-dead** [comment]
I am working on `rustc_const_eval`.

**Comment by nicklimmm** [comment]
I'm working on `rustdoc`, starting with `lib.rs`

**Comment by dayo05** [comment]
I'm going to work on` rustc_errors`

**Comment by abs0luty** [comment]
I'm going to work on `rustfmt`

**Comment by CraftSpider** [comment]
I'm picking up the apparently abandoned `rustc_hir_typeck`

**Comment by JeanCASPAR** [comment]
I'm taking `rustc_resolve`, which seems abandoned.

**Comment by jieyouxu** [comment]
> [!IMPORTANT]
> **Please see https://github.com/rust-lang/rust/issues/132181 for the current status of the compiler's translatable diagnostics infrastructure.**
